feat: optimistic mode

joanestebanr opened the pull request:
## Description
This PR add optimistic mode to aggkit. It checks if optimistic mode is active, if so, then instead of generating the FEP it call another end-point of `aggkit-prover` to generate a optimistic proof.

## Config: 
- Add new struct AggSender.OptimisticModeConfig:
   - `SovereignRollupAddr`: address of rollup, by default `AggSender.SovereignRollupAddr`
   - `TrustedSequencerKey`: key to sign optimistic certificates. Must match trustedSequencer and usually it's the same used for agglayer. So by default is `AggSender.AggsenderPrivateKey`
   - `OpNodeURL`: url to op-node
   - `RequireKeyMatchTrustedSequencer`: require that optmistic key match trustedKey (by default `true`)
  
- Define a new **defaut var** `AggsenderPrivateKey`  to allow override this key with  a KMS. It's still composed with `SequencerPrivateKeyPath` and `SequencerPrivateKeyPassword` so previous config file works in the same way.
- Disabled aggSender maxsize checked:  `AggSender.maxCertSize = 0`
- 

- Add new fields to DB (migration 0003): 
   - **extra_data**:  not structured data related to the cert for debug. e.g: optimistic store publicValues there
   - add new type `optimistic` to field **cert_type**

## Internal changes: 
- AggchainProverFlow: 
    -  configuration params have been grouped in a new struct `AggchainProverFlowConfig`
- Client for aggkit-prover have been moved from folder `aggsender/grpc` to `aggsender/aggchainproofclient`
- Moved variable `GenerateAggchainProofTimeout` from **MandatoryVars** (so it must be defined by the config-file) to **DefaultVars** that it defined a default variable so don't need to be defined on config-file
- Created a script `test/scipts/optimistic.sh` to allow check / modify contract status in kurtosis-cdk environment
- `baseFlow` object instead of a embedded object in the rest of flows is explicity call and use throught a interface to be able to mock in unittest
- Local debug scripts
      -  duplicated script to set a fep_optimistic debug: `local/local_config_fep_optimistic`and also created a specific config file for this case: `test/config/fep_optimistic.toml.template`
- Rendering confgfile allow to reference complex objects like the key configuration (it was required to allow `AggSender.OptimisticModeConfig.SovereignRollupAddr` = ` {{AggSender.SovereignRollupAddr}}`
Fixes #468


## Future tasks: 
- integrate e2e for optimistic

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: mocks, proof_generation_tool doesnt check optimisticMode

joanestebanr committed:
fix: ut

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: signature optimistic mode

joanestebanr committed:
feat: wip

joanestebanr committed:
fix: unittest

joanestebanr committed:
feat: wip

joanestebanr committed:
Merge branch 'develop' into feature/optimistic_mode_468

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: first full implementation, pending to fix ut and test it

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: first working version

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: coverage, unittest

goran-ethernal committed:
fix: failing UT

joanestebanr committed:
feat: fix ut

joanestebanr committed:
feat: local script by type of network

joanestebanr committed:
feat: convert baseFlow to interface

joanestebanr committed:
feat: config file specific for fep_optiimstic

joanestebanr committed:
feat: convert baseFlow to interface, and create unittest

joanestebanr committed:
feat: wip

joanestebanr committed:
feat: configuration key for optimistic mode point to aggsender one

joanestebanr committed:
Merge branch 'develop' into feature/optimistic_mode_468

joanestebanr committed:
fix: ut

joanestebanr committed:
fix: ut

joanestebanr committed:
fix: lint

joanestebanr committed:
fix: copilot comments

joanestebanr committed:
feat: increase coverage

joanestebanr committed:
feat: increase coverage

joanestebanr committed:
feat: increase coverage

joanestebanr committed:
feat: increase coverage

joanestebanr committed:
feat: increase coverage

joanestebanr committed:
feat: increase coverage

goran-ethernal commented on aggsender/flows/factory_test.go:
```diff
@@ -79,7 +101,17 @@ func TestNewFlow(t *testing.T) {
+				return 100, nil
+			}
 
+			mockL1Client.EXPECT().CallContract(mock.Anything, mock.Anything, mock.Anything).Return([]byte{1, 2, 3}, nil).Maybe()
```
Maybe you could do the same for `NewOptimisticModeQuerierFromContract` that you did for `var funcGetL2StartBlock = getL2StartBlock` and not have these mocks?

joanestebanr replied:
We have a huddle and agree a solution

joanestebanr replied:
We have agree that it's not required

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -51,31 +56,85 @@ func getL2StartBlock(sovereignRollupAddr common.Address, l1Client types.EthClien
 
 var funcNewEVMChainGERReader = chaingerreader.NewEVMChainGERReader
 
+type AggchainProverFlowConfig struct {
```
Add comments on the struct and constructor functions.

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -26,11 +25,17 @@ var errNoProofBuiltYet = &aggkitcommon.GRPCError{
+	aggchainProofClient   types.AggchainProofClientInterface
+	gerQuerier            types.GERQuerier
+	requireNoFEPBlockGap  bool
+	signer                signertypes.Signer
```
Should we rename this to `certificateSigner` perhaps. So that we know exactly what each signer do.

joanestebanr replied:
do you think that `certificateSigner` could work?

goran-ethernal replied:
Yes, that makes sense to me, since it is signing the certificate per se.

joanestebanr replied:
Done 053a3865ca48edba8d1664c49417c83ca246c89e

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -51,31 +56,85 @@ func getL2StartBlock(sovereignRollupAddr common.Address, l1Client types.EthClien
+
+// NewAggchainProverFlowInjectedBaseFlow returns a new instance of the AggchainProverFlow injecting baseFlow instead of
+// creating it
+func NewAggchainProverFlowInjectedBaseFlow(
```
If this is only used in tests, I would suggests to move this function to the test file, and make it a `test.Helper()`.

joanestebanr replied:
The idea would be deprecate `NewAggchainProverFlow` and start using this `NewAggchainProverFlowInjectedBaseFlow` because is more close to depency injections principle

goran-ethernal replied:
I do agree, but in your code, this constructor is only used in UTs. We agreed, that we will slice this BaseFlow to smaller interfaces that make more sense, and discard it all together. But, until then, if it is not used in real logic, I would suggest moving it to UT file.

joanestebanr replied:
I'm going to change the constructor for use this one

joanestebanr replied:
We agree to remove the old constructor in favour of the new one with injection of baseFlow

joanestebanr replied:
done c1004609cc217644782dd4d4302a6c75b05198b1

joanestebanr committed:
feat: increase coverage

joanestebanr committed:
feat: increase coverage

goran-ethernal commented on aggsender/flows/factory.go:
```diff
@@ -62,28 +66,39 @@ func NewFlow(
-			cfg.MaxCertSize,
-			startL2Block,
+			AggchainProverFlowConfig{
+				baseFlowConfig: BaseFlowConfig{
```
If it has only two config params, I am not sure if it is worth it to have the `BaseFlowConfig` 😕  . It does make sense to have a config struct for the `AggchainProverFlow`, and then use the params from that to pass it to the base flow.

joanestebanr replied:
Yes, I agree, that now it's a bit artificial, but when inject the baseFlow into the object it's going to own `baseFlowConfig` the idea of be an struct is that we can add parameters without the painful of changing an extra param everywhere (that the reason of the Default constructor)

joanestebanr replied:
After chainging the constructor to the one that inject the baseFlow se can remove the field AggchainProverFlowConfig. baseFlowConfig because will be no longer in use

joanestebanr replied:
done c1004609cc217644782dd4d4302a6c75b05198b1

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -97,11 +156,11 @@ func (a *AggchainProverFlow) sanityCheckNoBlockGaps(lastSentCertificate *types.C
 	}
 	msg := fmt.Sprintf("aggchainProverFlow - sanityCheckNoBlockGaps - last sent certificate: %s, startL2Block:%d",
-		lastSentCertficateStr, a.startL2Block)
+		lastSentCertficateStr, a.baseFlow.StartL2Block())
```
I am not sure why `baseFlow` needs this `StartL2Block`. It is only needed for the sanity check on the `aggchain prover flow`.

joanestebanr replied:
This variable was originally there, I see that baseFlow use it. We can talk about that and remove from baseFlow, for me it's also really artificial because this value is set by AggchainFlow

goran-ethernal replied:
Ahh, yes, sorry, I was only looking of references to the config field, not the function. I will resolve the comment.

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -171,8 +250,14 @@ func (a *AggchainProverFlow) GetCertificateBuildParams(ctx context.Context) (*ty
 
 		return buildParams, nil
 	}
+	// This line is just for emit a warning
```
```suggestion
	// This line is just for emitting a warning
```

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -300,14 +385,14 @@ func adjustBlockRange(buildParams *types.CertificateBuildParams,
+	certBuildParams *types.CertificateBuildParams,
 ) (*types.AggchainProof, *treetypes.Root, error) {
-	proof, leaf, root, err := a.l1InfoTreeDataQuerier.GetFinalizedL1InfoTreeData(ctx)
+	proof, leaf, root, err := a.baseFlow.L1InfoTreeDataQuerier().GetFinalizedL1InfoTreeData(ctx)
```
Hmmm, I am not sure about having this `baseFlow.GetInterface().Function()`. This makes mocking in the UTs even harder.

How about just having `GetFinalizedL1InfoTreeData` function on the `baseFlow` and just returning:
```go
func (b *baseFlow) GetFinalizedL1InfoTreeData() (*data, error) {
   return b.l1InfoTreeQuerier.GetFinalizedL1InfoTreeData()
 }
```
That way, in the test, you won't have to create two mocks, but just one. WDYT? Same goes for the `L2BridgeQuerier()`.

Then, after we merge this, we can separate the `baseFlow` into smaller interfaces that make more sense, and move away from this `base` class, to call it like that.

And once this gets merged, we can discuss

joanestebanr replied:
Yes, I dislike it, but the idea was to reduce the impact introducing the interface. Another option is that AggchainFlow has his own instance of whatever use (like  `l1InfoTreeQuerier`)  that for me, make more sense

goran-ethernal replied:
That would be fine as well. `aggchainProverFlow` having a field of its own, called `l1infotreeQuerier` and `l2bridgeQuerier`, and just use them.

joanestebanr replied:
Done c1004609cc217644782dd4d4302a6c75b05198b1

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -323,42 +408,47 @@ func (a *AggchainProverFlow) GenerateAggchainProof(
+	if !optimisticMode {
+		aggchainProof, err = a.aggchainProofClient.GenerateAggchainProof(request)
+	} else {
+		newLER, err := a.baseFlow.GetNewLocalExitRoot(ctx, certBuildParams)
```
This else looks like it belongs to some separate interface or a struct, called `optimistic_proof_query`, something like that. WDYT? It will probably make it easier to unit test the `flow_aggchain_prover`.

joanestebanr replied:
I don't get your idea... what do you mean: move the optimisticMode 'else' to another function?

goran-ethernal replied:
Yes, to move this logic in the else part to another function or even another file in your `optimistic` package, called `optimistic_proof_query` (this would make unit testing easier for `aggchainProverFlow`).

joanestebanr replied:
We agree to move this 'else' part to another function

joanestebanr replied:
done 2a67f21c04acb6fab2d4625b165b0dd517b3a8fb

joanestebanr committed:
fix: ut race condition

joanestebanr committed:
fix: lint and PR comments

joanestebanr committed:
- Refactor flows objects:
     - BaseFlow: Remove L1InfoTreeDataQuerier() and L2BridgeQuerier()
     - AggchainProverFlow: removed baseFlowConfig from wConfig
     - AggchainProverFlow: NewAggchainProverFlow injects baseFlow
     - PPFlow: constructor inject baseFlow

joanestebanr committed:
fix: ut

joanestebanr committed:
feat: PR comment, rename signer to certificateSigner

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -51,31 +58,55 @@ func getL2StartBlock(sovereignRollupAddr common.Address, l1Client types.EthClien
-	aggkitProverClient grpc.AggchainProofClientInterface,
+// AggchainProverFlowConfig holds the configuration for the AggchainProverFlow
+type AggchainProverFlowConfig struct {
+	requireNoFEPBlockGap bool
```
Do we need this now, since its one parameter only? If we need more parameters in the future, we can add a config struct then.

joanestebanr replied:
I prefer in that way, is going to be more simple to add new parameters (that for sure we are going to need soon)

goran-ethernal commented on aggsender/optimistic/optimistic.go:
```diff
@@ -0,0 +1,28 @@
+	"github.com/agglayer/aggkit/log"
+)
+
+func NewOptimistic(ctx context.Context,
```
Add a comment to the function.

goran-ethernal commented on aggsender/optimistic/optimistic_aggegation_proof_public_values_query.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+// The real object that implements OpNodeClienter is opnode.OpNodeClient
+var _ OpNodeClienter = (*opnode.OpNodeClient)(nil)
```
```suggestion
var (
      _ OpNodeClienter = (*opnode.OpNodeClient)(nil)
      _ FEPContractQuerier = (*aggchainfep.Aggchainfep)(nil)
      _ OptimisticAggregationProofPublicValuesQuerier = (*OptimisticAggregationProofPublicValuesQuery)(nil)
)
      
```

goran-ethernal commented on aggsender/optimistic/optimistic_aggegation_proof_public_values_query.go:
```diff
@@ -0,0 +1,77 @@
+
+var _ OptimisticAggregationProofPublicValuesQuerier = (*OptimisticAggregationProofPublicValuesQuery)(nil)
+
+type OptimisticAggregationProofPublicValuesQuery struct {
```
Add comments to struct an functions that are missing it.

goran-ethernal commented on aggsender/optimistic/optimistic_aggegation_proof_public_values_query.go:
```diff
@@ -0,0 +1,77 @@
+	}
+}
+
+// The parametameters are contained in the AggchainProofRequest struct
```
I think the comment is missing that describes what this function does.

goran-ethernal commented on aggsender/optimistic/optimistic_mode_query.go:
```diff
@@ -0,0 +1,36 @@
+type OptimisticModeQuerierFromContract struct {
+	aggchainFEPContract FEPContractQuerier
+	aggchainFEPAddr     common.Address
+	// op-node client
```
Remove this comment?

goran-ethernal commented on aggsender/optimistic/optimistic_sign.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+type OptimisticSignatureCalculatorImpl struct {
+	QueryAggregationProofPublicValues OptimisticAggregationProofPublicValuesQuerier
```
It seems that you don't need to export the fields on this struct. They can be private.

joanestebanr committed:
feat: PR comment move to another function requesting optimistic proof

goran-ethernal commented on aggsender/optimistic/optimistic_sign.go:
```diff
@@ -0,0 +1,118 @@
+func (o *OptimisticSignatureCalculatorImpl) Sign(ctx context.Context,
+	aggchainReq types.AggchainProofRequest,
+	newLocalExitRoot common.Hash,
+	certBuildParams *types.CertificateBuildParams,
```
Looks like you only need `Claims` from the certificate. Maybe you can pass just them, instead of build parameters object. WDYT?

goran-ethernal commented on aggsender/optimistic/optimistic_sign.go:
```diff
@@ -0,0 +1,118 @@
+		return nil, "", fmt.Errorf("aggregationProofPublicValues.Hash: error hashing aggregationProofPublicValues: %w", err)
+	}
+	importedBridgesHash := optimistichash.CalculateCommitImportedBrdigeExitsHashFromClaims(certBuildParams.Claims)
+	o.Logger.Infof("OptimisticSignatureCalculatorImpl.Sign aggHash:%s", aggregationProofPublicValuesHash.Hex())
```
Could we maybe merge these three log lines into one, and put it at the end when we sign the hash, or if signing the hash fails?

joanestebanr replied:
Done bac43c0756c4f63249d23127af6f70c5d359dfe3

joanestebanr committed:
fix: lint

goran-ethernal commented on aggsender/prover/proof_generation_tool.go:
```diff
@@ -82,19 +87,29 @@ func NewAggchainProofGenerationTool(
 	}
 
 	l1InfoTreeQuerier := query.NewL1InfoTreeDataQuerier(l1Client, l1InfoTreeSyncer)
+	// TODO: the signer it's required?
```
You can remove this comment. The signer is not required, since this tool is only used for regular requests to generate `aggchain proof`. We can add support for optimistic proofs later if its needed, but for now, you can remove the TODO.

joanestebanr replied:
I mean the certificateSigner, if you think that is not required I will deleted the TODO comment

goran-ethernal replied:
Yes. Its not required.

joanestebanr committed:
fix: ut

joanestebanr committed:
Merge branch 'develop' into feature/optimistic_mode_468

joanestebanr committed:
feat: PR comments improve log

joanestebanr committed:
feat: PR comments

joanestebanr committed:
feat: PR comments

joanestebanr committed:
feat: PR comments

joanestebanr committed:
feat: PR comments

joanestebanr committed:
feat: pr comments, change optimisticSign params

joanestebanr committed:
default maxcertlimit to same value as develop branch

joanestebanr committed:
fix: UT

temaniarpit27 commented on aggsender/db/migrations/0003.sql:
```diff
@@ -1,11 +1,15 @@
 -- +migrate Down
 ALTER TABLE certificate_info DROP COLUMN cert_type;
 ALTER TABLE certificate_info DROP COLUMN cert_source;
+ALTER TABLE certificate_info DROP COLUMN extra_data;
```
i hope that this migration hasn't gone in v0.3

joanestebanr replied:
No, this migration have been introduce in develop a few days ago, for that reason I take advatange of that instead of creating `0004.sql`

temaniarpit27 commented on aggsender/optimistic/optimistic_sign.go:
```diff
@@ -0,0 +1,123 @@
+)
+
+// OptimisticSignatureCalculator is an interface that defines the method for signing optimistic aggregation proofs.
+// type OptimisticSignatureCalculator interface {
```
commented code?

joanestebanr replied:
oops!

joanestebanr committed:
Merge branch 'develop' into feature/optimistic_mode_468

goran-ethernal commented on aggsender/flows/flow_aggchain_prover.go:
```diff
@@ -323,67 +380,85 @@ func (a *AggchainProverFlow) GenerateAggchainProof(
+		err := fmt.Errorf("aggchainProverFlow - error fetching aggchain proof (optimisticMode: %t) for lastProvenBlock: %d, "+
+			"maxEndBlock: %d. Err: %w. Message sent: %s", optimisticMode, lastProvenBlock, toBlock, err, request.String(),
 		)
+		a.log.Error(err.Error())
```
Do we need this log now? You are using it in the above lines to log it anyway.

joanestebanr replied:
Do you mean that is going to be logged eventually because we are returning as error? I think that is good to have the logs there because we can use the backtrace

goran-ethernal replied:
Ah, you are shadowing the err object, to create a new one to log it. Which is fine, I guess.

We need to review how we are passing and handling errors in the aggsender in future versions, because we have a lot of cases where we have something like this:

FunctionC() - catches an error, logs it, wraps it to a new error, and returns it.
FunctionB() - gets the wrapped error from FunctionC, logs it, wraps it and returns it.
FunctionA() - gets the wrapped error from FunctionB, logs it. And now, we have errors that are logged and wrapped multiple times.

But, this is something for the `0.6.0` version.

temaniarpit27 commented on aggsender/types/interfaces.go:
```diff
@@ -27,6 +27,21 @@ type AggsenderFlow interface {
 		buildParams *CertificateBuildParams) (*agglayertypes.Certificate, error)
 }
 
+type AggsenderFlowBaser interface {
```
baser? or base?

joanestebanr replied:
Baser... because it's an interface 😅

joanestebanr committed:
feat: render configfile can copy complex structures (like keys)

joanestebanr committed:
feat: add AggSender.OptimisticModeConfig.RequireKeyMatchTrustedSequencer to force sanity check of signing addr

joanestebanr committed:
fix: PR comments

joanestebanr committed:
test: disable RequireKeyMatchTrustedSequencer to see that pass CI

joanestebanr committed:
fix: ut

temaniarpit27 reviewed (approved):
LGTM

rachit77 commented on aggsender/flows/flow_base_test.go:
```diff
@@ -236,3 +239,135 @@ func Test_baseFlow_getNewLocalExitRoot(t *testing.T) {
+		nil,
+		NewBaseFlowConfigDefault())
+
+	t.Run("we can't generate new cert because a peding cert (not inError not Settle) we don't known if is going to settle or not ",
```
```suggestion
	t.Run("cannot generate new certificate because a pending certificate (neither InError nor Settled) has an unknown settlement status",
```

how about this?

joanestebanr replied:
Done 460cbb04e34ec8475ad9115659468620df9d6c41

joanestebanr replied:
Done 460cbb04e34ec8475ad9115659468620df9d6c41

joanestebanr committed:
feat: enable by default sanity check of trustedKey, to pass e2e create a config file to be use with this flag to false

joanestebanr committed:
feat: check kurtosis service are alive before launch bats tests

joanestebanr committed:
fix: e2e

joanestebanr committed:
fix: e2e aggkit-prover url

joanestebanr committed:
fix: e2e aggkit-prover url

joanestebanr committed:
fix: e2e aggkit-prover url
